internal/cmd/template_pluginpreset.go

```diff
@@ -0,0 +1,283 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+
+	helminternal "github.com/cloudoperators/greenhouse/internal/helm"
+
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/api/v1alpha1"
+)
+
+func init() {
+	templateCmd.AddCommand(newTemplatePluginPresetCmd())
+}
+
+type templatePluginPresetOptions struct {
+	pluginPresetPath     string
+	pluginDefinitionPath string
+	clusterName          string
+
+	pluginPreset            *greenhousev1alpha1.PluginPreset
+	clusterPluginDefinition *greenhousev1alpha1.ClusterPluginDefinition
+	values                  []greenhousev1alpha1.PluginOptionValue
+}
+
+func newTemplatePluginPresetCmd() *cobra.Command {
+	o := &templatePluginPresetOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "pluginpreset",
+		Short: "Template a ClusterPluginDefinition for a particular cluster",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.validate(); err != nil {
+				return err
+			}
+
+			if err := o.complete(); err != nil {
+				return err
+			}
+
+			return o.run()
+		},
+	}
+
+	cmd.Flags().StringVar(&o.pluginPresetPath, "pluginpreset", "", "Path to PluginPreset YAML file (required)")
+	cmd.Flags().StringVar(&o.pluginDefinitionPath, "plugindefinition", "", "Path to ClusterPluginDefinition YAML file (required)")
+	cmd.Flags().StringVar(&o.clusterName, "cluster", "", "Cluster name (required)")
+
+	if err := cmd.MarkFlagRequired("pluginpreset"); err != nil {
+		setupLog.Error(err, "Flag could not set as required", "pluginpreset")
+	}
+
+	if err := cmd.MarkFlagRequired("plugindefinition"); err != nil {
+		setupLog.Error(err, "Flag could not set as required", "plugindefinition")
+	}
+
+	if err := cmd.MarkFlagRequired("cluster"); err != nil {
+		setupLog.Error(err, "Flag could not set as required", "cluster")
+	}
+
+	return cmd
+}
+
+func (o *templatePluginPresetOptions) validate() error {
+	if _, err := os.Stat(o.pluginPresetPath); os.IsNotExist(err) {
+		return fmt.Errorf("PluginPreset file does not exist: %s", o.pluginPresetPath)
+	}
+
+	if _, err := os.Stat(o.pluginDefinitionPath); os.IsNotExist(err) {
+		return fmt.Errorf("ClusterPluginDefinition file does not exist: %s", o.pluginDefinitionPath)
+	}
+
+	return nil
+}
+
+func (o *templatePluginPresetOptions) complete() error {
+	var err error
+	o.clusterPluginDefinition, err = loadFromFile[greenhousev1alpha1.ClusterPluginDefinition](o.pluginDefinitionPath)
+	if err != nil {
+		return err
+	}
+
+	o.pluginPreset, err = loadFromFile[greenhousev1alpha1.PluginPreset](o.pluginPresetPath)
+	if err != nil {
+		return err
+	}
+
+	if err := o.validateCompatibility(); err != nil {
+		return err
+	}
```

Comment:
This does only work if the supplied files actually contain the correct kinds. The `loadFromFile` will also load a file containing a PluginDefinition into a PluginPreset struct.
The benefit is that it by coincidence works for both ClusterPluginDefinition and PluginDefinition, as both use the same struct for the `.spec`.

I think it is fine to leave the loading as is, but check in the `validateCompatibility` that the passed kinds are correct.